thunder/transforms/quantization.py

```diff
@@ -294,3 +295,214 @@ def transform_traces_pre_prologue(self, prologue_trace, computation_trace, epilo
 
         new_computation_trace.set_provenance(thunder.core.trace.TraceProvenance("quant pass"))
         return prologue_trace, new_computation_trace, epilogue_trace
+
+
+class LORATransform(Transform):
+    def __init__(
+        self,
+        r: int = 0,
+        lora_alpha: int = 1,
+        lora_dropout: float = 0.0,
+        **kwargs,
+    ):
+        self.r = r
+        self.lora_alpha = lora_alpha
+        self.lora_dropout = lora_dropout
+        self.lora_linear_names = set()
+        self.lora_linear_states = {}
+
+    def transform_module(self, model: thunder.ThunderModule):
+        self.thunder_module = model
+        shared_names = model._get_shared_names()
+        processed_names = set()
+
+        def convert_linear_submodule(tm, name, r):
+            self.lora_linear_names.add(name)
+            weight_name = f"{name}.weight"
+            processed_copies = shared_names[weight_name] & processed_names
+            if processed_copies:
+                copy_name = next(iter(processed_copies))
+                self.lora_linear_states[weight_name] = self.lora_linear_states[copy_name]
+                tm._overrides_parameters[weight_name] = tm._overrides_parameters[copy_name]
+                tm._overrides_parameters[f"{weight_name}.loraA"] = tm._overrides_parameters[f"{copy_name}.loraA"]
+                tm._overrides_parameters[f"{weight_name}.loraB"] = tm._overrides_parameters[f"{copy_name}.loraB"]
+                processed_names.add(weight_name)
+                return
+
+            w = tm.get_parameter(weight_name)
+            in_features, out_features = w.shape[0], w.shape[1]
+
+            loraA = torch.nn.Parameter(torch.empty((r, in_features)))
+            torch.nn.init.kaiming_uniform_(loraA, a=math.sqrt(5))
```

Comment:
I think we might move the initialization to a utility function as we will also need to do this when transforming the state dict.